service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/PlacementController.java

```diff
@@ -0,0 +1,110 @@
+package uk.gov.hmcts.reform.fpl.controllers;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.swagger.annotations.Api;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import uk.gov.hmcts.reform.ccd.client.model.AboutToStartOrSubmitCallbackResponse;
+import uk.gov.hmcts.reform.ccd.client.model.CallbackRequest;
+import uk.gov.hmcts.reform.ccd.client.model.CaseDetails;
+import uk.gov.hmcts.reform.fpl.enums.YesNo;
+import uk.gov.hmcts.reform.fpl.model.CaseData;
+import uk.gov.hmcts.reform.fpl.model.Child;
+import uk.gov.hmcts.reform.fpl.model.Placement;
+import uk.gov.hmcts.reform.fpl.model.common.Element;
+import uk.gov.hmcts.reform.fpl.model.common.dynamic.DynamicList;
+import uk.gov.hmcts.reform.fpl.service.PlacementService;
+
+import java.util.UUID;
+
+@Api
+@RestController
+@RequestMapping("/callback/placement")
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class PlacementController {
+
+    private final ObjectMapper mapper;
+    private final PlacementService placementService;
+
+    @PostMapping("/about-to-start")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToStart(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        boolean singleChild = placementService.hasSingleChild(caseData);
+
+        caseDetails.getData().put("singleChild", YesNo.from(singleChild));
+
+        if (singleChild) {
+            Element<Child> child = caseData.getAllChildren().get(0);
+            caseDetails.getData().put("placement", placementService.getPlacement(caseData, child));
+            caseDetails.getData().put("placementChildName", child.getValue().getParty().getFullName());
+        } else {
+            caseDetails.getData().put("childrenList", placementService.getChildrenList(caseData, null));
+        }
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .build();
+    }
+
+    @PostMapping("/mid-event")
+    public AboutToStartOrSubmitCallbackResponse handleMidEvent(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        UUID childId = getSelectedChildId(caseDetails, caseData);
+        Element<Child> child = placementService.getChild(caseData, childId);
+
+        caseDetails.getData().put("childrenList", placementService.getChildrenList(caseData, child));
```

Comment:
Other possibility is to use `caseDetails.getData().putAll(Map.of(...))`. It creates unnecessary `Map` instance, but makes it more readable. Thoughts?

Reply:
TBH, i dont see big benefits, still need to add 3 entries to map, and then map to map. Personally i think that creating another nesting does not improve readability.